src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,642 @@
-        const bit = set.free.findFirstSet() orelse unreachable;
-        set.free.unset(bit);
+        return set.try_acquire() orelse unreachable;
+    }
+
+    fn try_acquire(set: *BlockFreeSet) ?u64 {
+        if (set.try_acquire_recent()) |recent| return recent;
+
+        const index_bit = set.index.findFirstSet() orelse return null;
+        const shard_start = index_bit * shard_size;
+        const shard_end = std.math.min(shard_start + shard_size, set.blocks.bit_length);
+        assert(shard_start < set.blocks.bit_length);
+
+        const bit = findFirstSetBit(set.blocks, shard_start, shard_end) orelse return null;
+        assert(set.blocks.isSet(bit));
+
+        set.blocks.unset(bit);
+        // Update the index when every block in the shard is allocated.
+        if (findFirstSetBit(set.blocks, shard_start, shard_end) == null) {
+            set.index.unset(index_bit);
+        }
+
         const address = bit + 1;
         return @intCast(u64, address);
     }
 
+    fn try_acquire_recent(set: *BlockFreeSet) ?u64 {
+        const bit = set.recent.pop() orelse return null;
+        assert(set.blocks.isSet(bit));
+        set.blocks.unset(bit);
+        const shard = @divTrunc(bit, shard_size);
+        const shard_start = shard * shard_size;
+        const shard_end = std.math.min(shard_start + shard_size, set.blocks.bit_length);
+        if (findFirstSetBit(set.blocks, shard_start, shard_end) == null) set.index.unset(shard);
+        return @intCast(u64, bit + 1);
+    }
+
+    fn isFree(set: *BlockFreeSet, address: u64) bool {
+        return set.blocks.isSet(address - 1);
+    }
+
+    // Mark the specified block as free.
     pub fn release(set: *BlockFreeSet, address: u64) void {
         const bit = address - 1;
-        assert(!set.free.isSet(bit));
-        set.free.set(bit);
+        assert(!set.blocks.isSet(bit));
+
+        set.blocks.set(bit);
+        set.index.set(@divTrunc(bit, shard_size));
+        set.recent.push(bit) catch {};
+    }
+
+    pub fn decode(data: []const u8, allocator: *mem.Allocator) !BlockFreeSet {
+        var blocks = try BitsetEncoder(.immutable).decode(data, allocator);
+        errdefer blocks.deinit(allocator);
+
+        const index_size = divCeil(usize, blocks.bit_length, shard_size);
+        var index = try DynamicBitSetUnmanaged.initFull(index_size, allocator);
+        errdefer index.deinit(allocator);
+
+        var i: usize = 0;
+        while (i < index_size) : (i += 1) {
+            const shard_start = i * shard_size;
+            const shard_end = std.math.min(shard_start + shard_size, blocks.bit_length);
+            if (findFirstSetBit(blocks, shard_start, shard_end) == null) index.unset(i);
+        }
+
+        return BlockFreeSet{
+            .index = index,
+            .blocks = blocks,
+        };
+    }
+
+    // Returns the number of bytes that the `BlockFreeSet` needs to encode to.
+    pub fn encodeSize(set: BlockFreeSet) usize {
```

Comment:
I'm confused: right now it is (or at least should be) the exact number of bytes that the `BlockFreeSet` will encode to. That way the caller doesn't allocate any excess space for the `BlockFreeSet`.

Reply:
On second thought, we probably don't need this method. Because of static allocation, we actually always have a max-sized buffer available to encode into (regardless of how much the encoding ends up being).